File: NCPC/2014/ncpc2014/dicegame/submissions/accepted/lukas.cs
//Solution by lukasP (Lukáš Poláček)
//Construct the whole probability distributions
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;

class Problem
{
    public static void Main ()
    {
        var bs = new BufferedStream(Console.OpenStandardInput());
        StreamReader reader = new StreamReader(bs);

        List<double> pr = Enumerable.Repeat(0.0, 210).ToList();
        double win = 0, tie = 0;
        for (int i = 0; i < 2; i++)
        {
            int[] x = reader.ReadLine().Split().Select(y => int.Parse(y)).ToArray();

            double sz = (x[1] - x[0] + 1) * (x[3] - x[2] + 1);
            for (int j = x[0]; j <= x[1]; j++)
                for (int l = x[2]; l <= x[3]; l++)
                {
                    if (i == 1)
                    {
                        win += pr[j + l - 1] / sz;
                        tie += (pr[j + l] - pr[j + l - 1]) / sz;
                    }
                    else
                        pr[j + l] += 1 / sz;
                }

            if (i == 0) for (int j = 1; j < pr.Count; j++)
                pr[j] += pr[j - 1];
        }

        double sum = win * 2 + tie;
        if (Math.Abs(sum - 1) < 1e-9)
            Console.WriteLine("Tie");
        else
           Console.WriteLine(sum > 1 ? "Emma" : "Gunnar");
    }
}

File: NCPC/2014/ncpc2014/dicegame/submissions/accepted/lukas.py
#!/usr/bin/env python
# -*- coding: utf-8
#Solution by lukasP (Luká¿ Polá¿ek)
from sys import stdin

pr = [0.0] * 210
win, tie = 0.0, 0.0

for i in range(2):
    x = list(map(int, stdin.readline().split()))

    sz = float(x[1] - x[0] + 1) * (x[3] - x[2] + 1)
    for j in range(x[0], x[1] + 1):
        for l in range(x[2], x[3] + 1):
            if i == 1:
                win += pr[j + l - 1] / sz
                tie += (pr[j + l] - pr[j + l - 1]) / sz
            else:
                pr[j + l] += 1 / sz

    if i == 0:
        for j in range(1, len(pr)):
            pr[j] += pr[j - 1]

suma = win * 2 + tie;
if abs(suma - 1) < 1e-9:
    print("Tie")
else:
    print("Emma" if suma > 1 else "Gunnar")
